sycl/doc/DynamicLinkingTestPlan.md

```diff
@@ -0,0 +1,179 @@
+- Application compiled with `per_kernel` and library compiled with `per_source`
+- Application compiled with `per_kernel` and library compiled with `off`
+- Application compiled with `per_kernel` and library compiled with `auto`
+
+- Application compiled with `per_source` and library compiled with `per_kernel`
+- Application compiled with `per_source` and library compiled with `per_source`
+- Application compiled with `per_source` and library compiled with `off`
+- Application compiled with `per_source` and library compiled with `auto`
+
+and etc. Total 16 combinations.
+
+## Tests
+
+All the tests consist of at least one main application and at least one shared
+library.
+For all the tests simple calculation is performed by the application's kernel(s)
+that used `SYCL_EXTERNAL` from the shared library. Example : addition of
+vectors.
+The result of the calculation must be checked on host.
+
+### Basic test with shared library
+
+- Several kernels and `SYCL_EXTERNAL` functions are defined in both library and
+  application.
+- At least one `SYCL_EXTERNAL` function defined in shared library is used from
+  the application's kernel.
+- At least one `SYCL_EXTERNAL` function defined in shared library is used in
+  kernel inside the library.
+- Defined kernels are submitted and ran.
+
+### Test with dependent shared libraries
+
+The test consists of the main application and **two** shared libraries libA and
+libB and performs the following:
+- Application defines kernel(s) with dependency on libA.
+- libA defines `SYCL_EXTERNAL` function(s) with dependency on libB.
+- libB defines `SYCL_EXTERNAL` function(s) without external dependencies.
+- Defined kernels are submitted and ran.
+
+### Tests with function pointers
+
+#### Variation 1
+
+- Application defines a function with `indirectly_callable` attribute.
+- Shared library defines host API that can receive function pointer as an
+  argument and runs a kernel that calls through passed function pointer.
+- Address of `indirectly_callable` function is taken and passed to shared
+  library's host API.
+
+#### Variation 2
+
+- Shared library defines a function with `indirectly_callable` attribute.
+- Shared library defines host API that can return a function pointer to
+  a function with `indirectly_callable` attribute.
+- Application used host API of shared library and calls through received
+  function pointer from a kernel.
+
+### Test with shared library without device code
+
+The test aimed to check that support for device code inside shared library
+doesn't break support for host-only shared library.
+
+- Shared library defines host-only non-`SYCL_EXTERNAL` function.
+- The function from shared library is used by host part of the application.
+- Application defines and runs a kernel with a simple calculation.
+
+### Tests with run-time loaded shared library
+
+Tests aimed to check that support for dynamic linking of device code doesn't
+break run-time loading of shared library.
+
+#### Variation 1
```

Comment:
One more idea: shared library contains device code and provides host API to launch a kernel. The library is loaded through `dlopen` and the host API is queried through `dlsym`. Then we check that we can correctly launch the kernel from a runtime loaded shared library

Reply:
It actually seems more like it tests regular c++ functionality together with regular SYCL functionality. We already have some basic testing for runtime loading mentioned here to check that we didn't broke it by dynamic linking support. But since there is no explicit dependency on dynamic linking in these tests (and the one you proposed), should we arrange them to a separate suite?

Reply:
Fair enough, you can ignore this comment then (or we can apply it later or even to another test plan)